app/utils/function.ts

```diff
@@ -16,23 +16,28 @@ export function debounce<T extends (...args: any) => any>(
 
 type ThrottledFunction<T extends (...args: any) => any> = (
   ...args: Parameters<T>
-) => ReturnType<T>
+) => Promise<ReturnType<T>>
 
-// fn will be executed if there has not been any previous call during a delay
+// fn will be executed if there has not been any previous call during a delay or if a previous call is not still executing
 export function throttle<T extends (...args: any) => any>(
   fn: T,
   delayInMs: number,
 ): ThrottledFunction<T> {
   let inThrottle: boolean
   let lastResult: ReturnType<T>
 
-  return function (this: any, ...args: any[]): ReturnType<T> {
+  return async function (this: any, ...args: any[]): Promise<ReturnType<T>> {
     const context = this
 
     if (!inThrottle) {
       inThrottle = true
-      setTimeout(() => (inThrottle = false), delayInMs)
-      lastResult = fn.apply(context, args)
+      // the function could be async and could take longer than the delay to execute
+      let mightBePromise : ReturnType<T> = fn.apply(context, args)
```

Comment:
i suggest you make another function asyncThrottle since the behavior is quite different